### PR TITLE
#2768: Edit member domains - [RJM]

### LINKED ISSUE #2768: Member Profile page - Domain Assignment Page (Editable)
### Issue description

When editing an invited or active members permissions, an admin can also choose to adjust which domains the user is assigned to. For domain assignment, we started with the first page where an admin would just see all the domains the user has management rights to or invitations to. Now we need to add functionality to the edit button and allow an admin to change the assigned/invited domains. 

### Acceptance criteria

* Figma design here - [edit only](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=7364-28936&t=bqgIgKGnGc7JTI1y-4)
- [x] clicking Edit domain assignments on the Domain Assignment page opens a new view that has all the domains in the orgs with checkboxes next to them and matches figma above
- [x] only users with the correct permissions should be able to view this page (are part of the org, and have the create/edit org member permission)
- [x] read and include include annotations 2-6  (these are summed up below)
- [x] The domains table is searchable with the search text being in EDIT state "search all domains" (this may be already implemented in a previous ticket and be fully reusable)
- [x] For editable domains, there should be 3 states: default, active, and disabled ([see Figma for reference](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8901-84338&t=bqgIgKGnGc7JTI1y-4)) Domains that are already assigned to the user will be in Active state (pre-check box'd)
- [x] a user can check and uncheck domain names when they aren't disabled.
- [x] Have sortability for domains that are checked and unchecked ([see Figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8938-28420&t=bqgIgKGnGc7JTI1y-4))
- [x] Have sortability on domain names ([see Figma](https://www.figma.com/design/xFtGLHVrhp0lvwh0gYWnbx/.gov-registar?node-id=8938-28420&t=bqgIgKGnGc7JTI1y-4))
- [x] Disable rows for already assigned domains per the annotation
- [x] Clicking "Cancel" navigates the user back to the "view only" page, and clicking "Review" button does not do anything (there will be a redirect to another page that is added in the next ticket #2853)

### Additional context

* Be sure to make sure the disabled text/row is screen reader friendly pls!

### Links to other issues
* Blocked by #2767 
* Blocks #2853

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #2768

## Changes

 
- Only users with the correct permissions should be able to view this page (are part of the org, and have the create/edit org member permission)
- Edit breadcrumbs - NOTE - The breadcrumbs reflect the router config, and Figma is missing the penultimate step/link of the readonly domains list
- Default sort is name is alphabetical 
- Initialize checkboxes
- A list tracking added domains on the frontend { id, name }
- A list tracking removed domains on the frontend { id, name }
- Ajax list of portfolio domains
- Ajax list of assigned domains from db and layer on top of the DOM draw
- Changes we are keeping track of and layer on top of the DOM draw
- Sole domain manager boolean is returned in the serialized JSON
- Pass in the added and removed lists of IDs into the ajax call
- Backend handles sortby checkbox. The domains show in alphabetical order
- The ANDI output should be Currently sorted by assigned domains.Click to unsort - NOTE - the uswds implementation is slightly different, when with that (it matches the pattern for all other sortable columns)
- Cancel navigates the user back to the "view only" page,
- Cancel triggers a modal if one of the 2 lists is not length 0
- Clean up item number indicator
- Set pagination size at a high number
- Unit test backend changes
- Unrelated to PR: Fix bug on search label on the view member domains list
- NOTE: modal 'go back' is left as 'cancel' matching other modals
- NOTE: hitting cancel when you have a selection will trigger a modal, navigating away from the page through other means (hitting anther link, clicking the browser back button) will not trigger a modal. This is because the cost to do so in a non SPA is high, and in the case of the back button it goes into shady practice territory.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Update documentation in READMEs and/or onboarding guide

#### Ensured code standards are met (Original Developer)

- [x] If any updated dependencies on Pipfile, also update dependencies in requirements.txt.
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [x] Tag @dotgov-designers in this PR's Reviewers for design review. If code is not user-facing, delete design reviewer checklist
- [x] Verify new pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [x] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Verified code meets all checks above. Address any checks that are not satisfied
- [ ] Reviewed this code and left comments. Indicate if comments must be addressed before code is merged
- [ ] Checked that all code is adequately covered by tests
- [ ] Verify migrations are valid and do not conflict with existing migrations

#### Validated user-facing changes as a developer
**Note:** Multiple code reviewers can share the checklists above, a second reviewer should not make a duplicate checklist. All checks should be checked before approving, even those labeled N/A. 

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] (Rarely needed) Tested as both an analyst and applicant user

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior. Comment any found issues or broken flows.
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari
- [ ] (Rarely needed) Tested as both an analyst and applicant user

### References
- [Code review best practices](../docs/dev-practices/code_review.md)

### CODE REVIEW

File: src/registrar/context_processors.py [hunk 1]
```diff
@@ -108,10 +108,20 @@ def is_widescreen_mode(request):
         "/no-organization-domains/",
         "/domain-request/",
     ]
+    exclude_paths = [
```
Comment (rachidatecs) [inline src/registrar/context_processors.py:111]: widescreen_paths can be a bear as it trickles down sub-urls. exclude_paths gives us a way out.

**Comment by Matt-Spence** [comment]
When I refresh the "edit domain assignments" page, the "continue without saving" modal pops up until the table populates. I can't get a screenshot because its only there for like half a second but it should be an easy behavior to replicate.

**Comment by rachidatecs** [comment]
> When I refresh the "edit domain assignments" page, the "continue without saving" modal pops up until the table populates. I can't get a screenshot because its only there for like half a second but it should be an easy behavior to replicate.

I'll take a look, but I bet I know what that is: When you write markup for a USWDS modal, it's just a bunch of content that sits in the middle of your page until USWDS JS rips it out and injects its functional markup at the end of the DOM. The flash you may be seeing is that initial markup if the USWDS JS is slow to load and execute. Not much we can do about that, unfortunately.

I'll double check anyways: We can put a breakpoint on the JS which will prevent the code from running, so the page will be stuck in pre-init mode. We can then confirm that this is what you are seeing.

**Comment by rachidatecs** [comment]
@Matt-Spence: 

Screenshot 2024-12-11 at 12 31 56 PM

**Comment by Matt-Spence** [comment]
> @Matt-Spence:
> 
> Screenshot 2024-12-11 at 12 31 56 PM

Yep this is it. Is this expected/acceptable behavior? Just struck me as odd.

File: src/registrar/assets/src/js/getgov/table-edit-member-domains.js [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+   * Initializes listeners on checkboxes in the table. Checkbox listeners are used
+   * in this case to track changes to domain assignments in js (addedDomains and removedDomains)
+   * before changes are saved.
+   * initCheckboxListeners is called each time table is loaded.
+   */
+  initCheckboxListeners() {
+    const checkboxes = this.tableWrapper.querySelectorAll('input[type="checkbox"]');
+    checkboxes.forEach(checkbox => {
+      checkbox.addEventListener('change', () => {
+        const domain = { id: +checkbox.value, name: checkbox.name };
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/assets/src/js/getgov/table-edit-member-domains.js:157]: The plus sign turns string into an int

**Review by Matt-Spence — APPROVED** [review]
LGTM

File: src/registrar/views/member_domains_json.py [hunk 3]
```diff
@@ -107,9 +154,22 @@ def serialize_domain(self, domain, user):
...
+        # Check if the specified member is the only member assigned as manager of domain
+        only_member_assigned_to_domain = False
+        if member_id:
+            member_domain_role_count = UserDomainRole.objects.filter(
+                domain_id=domain.id, role=UserDomainRole.Roles.MANAGER
+            ).count()
+            member_domain_role_exists = UserDomainRole.objects.filter(
+                domain_id=domain.id, user_id=member_id, role=UserDomainRole.Roles.MANAGER
+            ).exists()
+            only_member_assigned_to_domain = member_domain_role_exists and member_domain_role_count == 1
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/views/member_domains_json.py:167]: > This should be equivalent with one less db call, right? I think count just returns 0 when nothing exists

Count returns 0 when nothing exists, but these are two slightly different filters. One checks if the particular member has a role on the domain. The other checks the count of members who have roles on the domain. For only member assigned to domain to be True, the particular member must have a role on the domain, and they must be the only member with a role on the domain.

File: src/registrar/tests/test_views_member_domains_json.py [hunk 4]
```diff
@@ -70,8 +73,10 @@ def setUpClass(cls):
             roles=[UserPortfolioRoleChoices.ORGANIZATION_ADMIN],
         )
         # Add user_member as manager of domains
-        UserDomainRole.objects.create(user=cls.user_member, domain=cls.domain1)
-        UserDomainRole.objects.create(user=cls.user_member, domain=cls.domain2)
+        UserDomainRole.objects.create(user=cls.user_member, domain=cls.domain1, role=UserDomainRole.Roles.MANAGER)
+        UserDomainRole.objects.create(user=cls.user_member, domain=cls.domain2, role=UserDomainRole.Roles.MANAGER)
+        UserDomainRole.objects.create(user=cls.user_member, domain=cls.domain3, role=UserDomainRole.Roles.MANAGER)
+        UserDomainRole.objects.create(user=cls.user_no_perms, domain=cls.domain3, role=UserDomainRole.Roles.MANAGER)
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_views_member_domains_json.py:79]: No, this is to test two conditions, (1) when there are more than one role on a domain, and (2) when there are no roles on a domain. In this case, two users added to domain3 and none to domain4

**Pull request merged**